pkg/operator/ceph/csi/config.go

```diff
@@ -0,0 +1,88 @@
+/*
+Copyright 2024 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	"context"
+
+	cephclient "github.com/rook/rook/pkg/daemon/ceph/client"
+
+	csiopv1a1 "github.com/ceph/ceph-csi-operator/api/v1alpha1"
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func CreateUpdateClientProfileRadosNs(ctx context.Context, c client.Client, clusterInfo *cephclient.ClusterInfo, cephBlockPoolRadosNamespacedName types.NamespacedName, clusterID, clusterName string) error {
+
+	logger.Info("creating ceph-csi clientProfile cr for rados namespace")
+
+	csiOpClientProfile := &csiopv1a1.ClientProfile{}
+	csiOpClientProfile.Name = clusterID
+	csiOpClientProfile.Namespace = cephBlockPoolRadosNamespacedName.Namespace
+	csiOpClientProfile.Spec = csiopv1a1.ClientProfileSpec{
+		CephConnectionRef: v1.LocalObjectReference{
+			Name: clusterName,
+		},
+		Rbd: &csiopv1a1.RbdConfigSpec{
+			RadosNamespace: cephBlockPoolRadosNamespacedName.Name,
+		},
+	}
+
+	err := clusterInfo.OwnerInfo.SetOwnerReference(csiOpClientProfile)
+	if err != nil {
+		return errors.Wrapf(err, "failed to set owner reference on ClientProfile CR %q", csiOpClientProfile.Name)
+	}
+
+	err = c.Create(ctx, csiOpClientProfile)
+	if err != nil && !kerrors.IsAlreadyExists(err) {
+		return errors.Wrapf(err, "failed to create ceph-csi clientProfile cr for Rbd %s", csiOpClientProfile.Name)
+	}
+
+	return nil
+}
+
+func CreateUpdateClientProfileSubVolGrp(ctx context.Context, c client.Client, clusterInfo *cephclient.ClusterInfo, cephFilesystemNamespacedName types.NamespacedName, clusterID, clusterName string) error {
+
+	logger.Info("Creating ceph-csi clientProfile CR for subvolume group")
+
+	csiOpClientProfile := &csiopv1a1.ClientProfile{}
+	csiOpClientProfile.Name = clusterID
+	csiOpClientProfile.Namespace = cephFilesystemNamespacedName.Namespace
+	csiOpClientProfile.Spec = csiopv1a1.ClientProfileSpec{
+		CephConnectionRef: v1.LocalObjectReference{
+			Name: clusterName,
+		},
+		CephFs: &csiopv1a1.CephFsConfigSpec{
+			SubVolumeGroup: cephFilesystemNamespacedName.Name,
+		},
+	}
+
+	err := clusterInfo.OwnerInfo.SetOwnerReference(csiOpClientProfile)
+	if err != nil {
+		return errors.Wrapf(err, "failed to set owner reference on  ClientProfile CR %q", csiOpClientProfile.Name)
+	}
+
+	err = c.Create(ctx, csiOpClientProfile)
```

Comment:
> the spec will not changes that's why no update here

Perhaps today there are no fields to update, but settings could be added in the future. Even now, looks like https://github.com/ceph/ceph-csi-operator/pull/64 just added a couple of fields that could change. So I'd still recommend calling update on all the CRs.

Reply:
added